src/Integration/Persistence/FileBindingSerializer.cs

```diff
@@ -0,0 +1,230 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.Alm.Authentication;
+using Newtonsoft.Json;
+using SonarLint.VisualStudio.Integration.Helpers;
+using SonarLint.VisualStudio.Integration.Resources;
+using SonarQube.Client.Helpers;
+
+namespace SonarLint.VisualStudio.Integration.Persistence
+{
+    internal abstract class FileBindingSerializer : ISolutionBindingSerializer
+    {
+        private readonly ICredentialStore credentialStore;
+        private readonly IFile fileWrapper;
+
+        protected readonly ISourceControlledFileSystem sccFileSystem;
+        protected readonly ILogger logger;
+
+        protected enum WriteMode
+        {
+            // The file will be enqued but not actually written.
+            // It is the responsibility of the caller to flush the queue.
+            // This option should be used if multiple other files are being
+            // written using the same instance of the SCC wrapper
+            // (e.g. legacy connected mode, which also writes the rulesets)
+            Queued,
+
+            // The file will be written to the SCC file system immediately
+            Immediate
+        }
+
+        protected FileBindingSerializer(ISourceControlledFileSystem sccFileSystem, ICredentialStore store,
+            ILogger logger, IFile fileWrapper)
+        {
+            if (sccFileSystem == null)
+            {
+                throw new ArgumentNullException(nameof(sccFileSystem));
+            }
+            if (store == null)
+            {
+                throw new ArgumentNullException(nameof(store));
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            if (fileWrapper == null)
+            {
+                throw new ArgumentNullException(nameof(fileWrapper));
+            }
+
+            this.sccFileSystem = sccFileSystem;
+            this.credentialStore = store;
+            this.logger = logger;
+            this.fileWrapper = fileWrapper;
+        }
+
+        protected abstract WriteMode Mode { get; }
+
+        protected abstract string GetFullConfigurationFilePath();
+
+        protected virtual bool OnSuccessfulFileWrite(string filePath)
+        {
+            // Default implementation is a no-op
+            return true;
+        }
```

Comment:
Extension point in case the child class wants to do extra work after the file is written (the legacy serializer does - it adds the file to the solution). The new connected mode doesn't need to do anything.